remix-lib/src/execution/txRunner.js

```diff
@@ -102,43 +104,66 @@ class TxRunner {
       timestamp: timestamp,
       nonce: new BN(account.nonce++),
       gasPrice: new BN(1),
-      gasLimit: new BN(gasLimit, 10),
+      gasLimit: gasLimit,
       to: to,
       value: new BN(value, 10),
       data: Buffer.from(data.slice(2), 'hex')
     })
     tx.sign(account.privateKey)
 
-    const coinbases = [ '0x0e9281e9c6a0808672eaba6bd1220e144c9bb07a', '0x8945a1288dc78a6d8952a92c77aee6730b414778', '0x94d76e24f818426ae84aa404140e8d5f60e10e7e' ]
-    const difficulties = [ new BN('69762765929000', 10), new BN('70762765929000', 10), new BN('71762765929000', 10) ]
+    const coinbases = ['0x0e9281e9c6a0808672eaba6bd1220e144c9bb07a', '0x8945a1288dc78a6d8952a92c77aee6730b414778', '0x94d76e24f818426ae84aa404140e8d5f60e10e7e']
+    const difficulties = [new BN('69762765929000', 10), new BN('70762765929000', 10), new BN('71762765929000', 10)]
     var block = new EthJSBlock({
       header: {
         timestamp: timestamp || (new Date().getTime() / 1000 | 0),
         number: self.blockNumber,
         coinbase: coinbases[self.blockNumber % coinbases.length],
         difficulty: difficulties[self.blockNumber % difficulties.length],
-        gasLimit: new BN(gasLimit, 10).imuln(2)
+        gasLimit: new BN('8000000').imuln(1)
       },
-      transactions: [],
+      transactions: [tx],
       uncleHeaders: []
     })
     if (!useCall) {
       ++self.blockNumber
     } else {
-      executionContext.vm().stateManager.checkpoint(() => {})
+      executionContext.vm().stateManager.checkpoint(() => { })
     }
 
-    executionContext.vm().runTx({block: block, tx: tx, skipBalance: true, skipNonce: true}, function (err, result) {
-      if (useCall) {
-        executionContext.vm().stateManager.revert(function () {})
-      }
-      err = err ? err.message : err
-      if (result) {
-        result.status = '0x' + result.vm.exception.toString(16)
-      }
-      callback(err, {
-        result: result,
-        transactionHash: ethJSUtil.bufferToHex(Buffer.from(tx.hash()))
+    this.checkpointAndCommit(() => {
+      executionContext.vm().runBlock({ block: block, generate: true, skipBlockValidation: true, skipBalance: false }, function (err, results) {
+        err = err ? err.message : err
+        if (err) {
+          return callback(err)
+        }
+        let result = results.results[0]
+        if (useCall) {
+          executionContext.vm().stateManager.revert(function () { })
+        }
+        if (result) {
+          result.status = '0x' + result.vm.exception.toString(16)
+        }
+
+        executionContext.addBlock(block)
+        executionContext.trackTx('0x' + tx.hash().toString('hex'), block)
+
+        callback(err, {
+          result: result,
+          transactionHash: ethJSUtil.bufferToHex(Buffer.from(tx.hash()))
+        })
+      })
+    })
+  }
+
+  checkpointAndCommit (cb) {
+    if (executionContext.vm().stateManager._checkpointCount > 1) {
```

Comment:
not sure about the context, but could we remove the use of `_checkpointCount` here?